edgelet/edgelet-kube/src/module/authentication.rs

```diff
@@ -0,0 +1,118 @@
+use crate::{KubeModuleRuntime, KubeRuntimeData};
+
+pub fn authenticate<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    req: &Request<Body>,
+) -> impl Future<Item = AuthId, Error = Error>
+where
+    T: TokenSource + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    req.headers()
+        .typed_get::<Authorization>()
+        .map(|auth| {
+            auth.and_then(|auth| {
+                auth.as_bearer().map(|token| {
+                    let client_copy = runtime.client();
+                    let namespace = runtime.namespace().to_owned();
+                    let fut = runtime
+                        .client()
+                        .lock()
+                        .expect("Unexpected lock error")
+                        .borrow_mut()
+                        .token_review(runtime.namespace(), token.as_str())
+                        .map_err(|err| {
+                            log_failure(Level::Warn, &err);
+                            Error::from(err)
+                        })
+                        .and_then(move |token_review| {
+                            token_review
+                                .status
+                                .as_ref()
+                                .filter(|status| status.authenticated.filter(|x| *x).is_some())
+                                .and_then(|status| {
+                                    status.user.as_ref().and_then(|user| user.username.clone())
+                                })
+                                .map_or(Either::A(future::ok(AuthId::None)), |name| {
+                                    Either::B(get_module_original_name(
+                                        &client_copy,
+                                        &namespace,
+                                        &name,
+                                    ))
+                                })
+                        });
+
+                    Either::A(fut)
+                })
+            })
+            .unwrap_or_else(|| Either::B(future::ok(AuthId::None)))
+        })
+        .map_err(Error::from)
+        .into_future()
+        .flatten()
+}
+
+fn get_module_original_name<T, S>(
+    client: &Arc<Mutex<RefCell<KubeClient<T, S>>>>,
+    namespace: &str,
+    username: &str,
+) -> impl Future<Item = AuthId, Error = Error>
+where
+    T: TokenSource + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    match username.split(':').last() {
+        Some(name) => Either::A({
+            let name = name.to_owned();
+
+            client
+                .lock()
+                .expect("Unexpected lock error")
+                .borrow_mut()
+                .get_service_account(namespace, &name)
```

Comment:
Once again, does case matter at all here? Should we normalize things behind the scenes?

Reply:
`name` here is a part of the value that Kube API server responded to `token_review` call. Thus it already conforms to k8s standards. I extract only the last part of the full name and use it as-is as a parameter to get a service account description for the name.

Other words no normalization required because we use k8s provided values to get data from k8s.